xbmc/pvr/timers/PVRTimers.cpp

```diff
@@ -105,24 +105,38 @@ bool CPVRTimers::IsRecording(void) const
 
 bool CPVRTimers::SetEpgTagTimer(const CPVRTimerInfoTagPtr &timer)
 {
-  const CEpgInfoTagPtr epgTag(timer->GetEpgInfoTag());
-  if (epgTag)
-  {
-    epgTag->SetTimer(timer);
-    return true;
-  }
-  return false;
+  if (timer->m_bStartAnyTime || timer->m_bEndAnyTime)
+    return false;
+
+  std::vector<CEpgInfoTagPtr> tags(g_EpgContainer.GetEpgTagsForTimer(timer));
+
+  if (tags.empty())
+    return false;
+
+  // assign first matching epg tag to the timer.
+  timer->SetEpgTag(tags.front());
```

Comment:
<code>CPVRTimerInfoTag::GetEpgTag</code> is called a gazillion of times. Obtaining the tag on demand is quite expensive as the epg tag must be found among potentially tens of thousands of epg events. So, if you found it once, hold it (or set it from outside if you already have it).

On my long list of things to improve at some later point in time is removal of the circular dependency between CPVRTimerInfoTag and CEpgInfoTag instances, though. Tried it once, but it quickly turned out that this is not a no-brainer.